### PR TITLE
451006-RomanAlshevscky

### PR DESCRIPTION
**Pull request by rom707**
Ready

### CODE REVIEW

**Comment by RamanBut-Husaim** [comment]
@rom707 Configure .gitignore properly and remove all .exe, .pdb, etc. files.

**Comment by rom707** [comment]
Unnecessary files removed.

File: Tracer/TracedMethodItem.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Diagnostics;
+namespace TracerLab
+{
+    public class TracedMethodItem
+    {
+        internal readonly string name;
```
Comment (RamanBut-Husaim) [inline Tracer/TracedMethodItem.cs]: @rom707 Change to properties.

File: Tracer/TracerXmlFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            methodEl.Add(new XAttribute("class", item.className));
+            methodEl.Add(new XAttribute("name", item.name));
+            methodEl.Add(new XAttribute("arguments", item.argCount));
+            methodEl.Add(new XAttribute("time", item.timer.ElapsedMilliseconds));
+            return methodEl;
+        }
+
+        private void Save(XDocument xmlDocument, string path)
+        {
+           /* XmlWriterSettings settings = new XmlWriterSettings
```
Comment (RamanBut-Husaim) [inline Tracer/TracerXmlFormatter.cs]: @rom707 Remove `dead` code.

File: Tracer/TracerXmlFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using System.Xml;
+
+namespace TracerLab
+{
+    public class TracerXmlFormatter : ITraceFormatter
+    {
+        private string xmlFilePath;
```
Comment (RamanBut-Husaim) [inline Tracer/TracerXmlFormatter.cs]: @rom707 `readonly`?

File: Tracer/Tracer.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        private Tracer()
+        {
+            _traceResult = new TraceResult();
+        }
+
+        public void StartTrace()
+        {
+            int threadId = System.Threading.Thread.CurrentThread.ManagedThreadId;
+            int callDepth = 0;
+            if (!tracedThreads.ContainsKey(threadId))
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @rom707 You are modifying the collection in a non-thread manner.

File: Tracer/TraceResult.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TracerLab
+{
+    public class TraceResult
+    {
+        //id, callStack
+        public Dictionary<int, Stack<TracedMethodItem>> tracedThreads;
```
Comment (RamanBut-Husaim) [inline Tracer/TraceResult.cs]: @rom707 Public field. Replace with property.

File: Tracer/TracedMethodItem.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Diagnostics;
+namespace TracerLab
+{
+    public class TracedMethodItem
+    {
+        internal string name { get; set; }
```
Comment (RamanBut-Husaim) [inline Tracer/TracedMethodItem.cs]: @rom707 `PascalCase` for property naming.

File: Tracer/Tracer.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                lock (CollectionAccessLock)
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @rom707 Incorrect approach.

**Comment by RamanBut-Husaim** [comment]
@rom707 Accepted.

**Pull request merged**